src/aspire/orientation/__init__.py

```diff
@@ -1,38 +1,497 @@
+            idx = np.arange(4 * shift_eq_idx, 4 * shift_eq_idx + 4)
+            # angle of common ray in image i
+            shift_alpha = c_ij * d_theta
+            # Angle of common ray in image j.
+            shift_beta = c_ji * d_theta
+            # Row index to construct the sparse equations
+            shift_i[idx] = shift_eq_idx
+            # Columns of the shift variables that correspond to the current pair [i, j]
+            shift_j[idx] = [2 * i, 2 * i + 1, 2 * j, 2 * j + 1]
+            # Right hand side of the current equation
+            shift_b[shift_eq_idx] = dx
+
+            # Compute the coefficients of the current equation
+            coeffs = np.array([np.sin(shift_alpha), np.cos(shift_alpha),
+                               -np.sin(shift_beta), -np.cos(shift_beta)])
+            shift_eq[idx] = -1 * coeffs if is_pf_j_flipped else coeffs
+
+        # create sparse matrix object only containing non-zero elements
+        shift_equations = sparse.csr_matrix((shift_eq, (shift_i, shift_j)),
+                                            shape=(n_equations, 2 * n_img),
+                                            dtype=np.float64)
+
+        return shift_equations, shift_b
+
+    def _estimate_num_shift_equations(self, n_img, equations_factor=1, max_memory=4000):
         """
-        perform estimation of orientations
+        Estimate total number of shift equations in images
+
+        The function computes total number of shift equations based on
+        number of images and preselected memory factor.
+        :param n_img:  The total number of input images
+        :param equations_factor: The factor to rescale the number of shift equations
+            (=1 in default)
+        :param max_memory: If there are N images and N_check selected to check
+            for common lines, then the exact system of equations solved for the shifts
+            is of size 2N x N(N_check-1)/2 (2N unknowns and N(N_check-1)/2 equations).
+            This may be too big if N is large. The algorithm will use `equations_factor`
+            times the total number of equations if the resulting total number of
+            memory requirements is less than `max_memory` (in megabytes); otherwise it
+            will reduce the number of equations to fit in `max_memory`.
+        :return: Estimated number of shift equations
         """
-        pass
+        # Number of equations that will be used to estimation the shifts
+        n_equations_total = int(np.ceil(n_img * (self.n_check - 1) / 2))
+        # Estimated memory requirements for the full system of equation.
+        # This ignores the sparsity of the system, since backslash seems to
+        # ignore it.
+        memory_total = equations_factor * (
+                n_equations_total * 2 * n_img * np.dtype('float64').itemsize)
+        if memory_total < (max_memory * 10 ** 6):
+            n_equations = n_equations_total
```

Comment:
`np.ceil(equations_factor * n_equations_total)`

Reply:
Done in commit e1c9992.